File: 0423-reconstruct-original-digits-from-english/0423-reconstruct-original-digits-from-english.py
from collections import Counter

class Solution(object):
    def originalDigits(self, s):
        count = Counter(s)
        digits = [0] * 10

        # Digits with unique identifying letters
        digits[0] = count['z']  # zero
        digits[2] = count['w']  # two
        digits[4] = count['u']  # four
        digits[6] = count['x']  # six
        digits[8] = count['g']  # eight

        # Remaining digits
        digits[3] = count['h'] - digits[8]          # three
        digits[5] = count['f'] - digits[4]          # five
        digits[7] = count['s'] - digits[6]          # seven
        digits[1] = count['o'] - digits[0] - digits[2] - digits[4]
        digits[9] = count['i'] - digits[5] - digits[6] - digits[8]

        result = []

        for digit in range(10):
            result.append(str(digit) * digits[digit])

        return "".join(result)

File: 0423-reconstruct-original-digits-from-english/0423-reconstruct-original-digits-from-english.cpp
class Solution {
public:
    string originalDigits(string s) {
        vector<int> count(26, 0);

        // Count every character
        for (char c : s) {
            count[c - 'a']++;
        }

        vector<int> digit(10, 0);

        // Unique characters
        digit[0] = count['z' - 'a'];  // zero
        digit[2] = count['w' - 'a'];  // two
        digit[4] = count['u' - 'a'];  // four
        digit[6] = count['x' - 'a'];  // six
        digit[8] = count['g' - 'a'];  // eight

        // Remove the letters already used
        digit[3] = count['h' - 'a'] - digit[8];              // three
        digit[5] = count['f' - 'a'] - digit[4];              // five
        digit[7] = count['s' - 'a'] - digit[6];              // seven

        digit[1] = count['o' - 'a'] - digit[0]
                                  - digit[2]
                                  - digit[4];                // one

        digit[9] = count['i' - 'a'] - digit[5]
                                  - digit[6]
                                  - digit[8];                // nine

        // Build answer in ascending order
        string ans;

        for (int i = 0; i <= 9; i++) {
            while (digit[i] > 0) {
                ans += char('0' + i);
                digit[i]--;
            }
        }

        return ans;
    }
};